cloudinit/sources/helpers/vmware/imc/config_nic.py

```diff
@@ -87,161 +80,190 @@ def find_devices(self):
-            subnet.update({"type": "dhcp"})
-            return ([subnet], route_list)
+            dhcp4.update({"dhcp4": True})
+            # dhcp4-overrides
+            if self.name_servers or self.dns_suffixes:
+                dhcp4.update({"dhcp4-overrides": {"use-dns": False}})
         else:
-            subnet.update({"type": "static"})
-
-        # Static Ipv4
-        addrs = nic.staticIpv4
-        if not addrs:
-            return ([subnet], route_list)
-
-        v4 = addrs[0]
-        if v4.ip:
-            subnet.update({"address": v4.ip})
-        if v4.netmask:
-            subnet.update({"netmask": v4.netmask})
-
-        # Add the primary gateway
-        if nic.primary and v4.gateways:
-            self.ipv4PrimaryGateway = v4.gateways[0]
-            subnet.update({"gateway": self.ipv4PrimaryGateway})
-            return ([subnet], route_list)
-
-        # Add routes if there is no primary nic
-        if not self._primaryNic and v4.gateways:
-            subnet.update(
-                {"routes": self.gen_ipv4_route(nic, v4.gateways, v4.netmask)}
-            )
-
-        return ([subnet], route_list)
-
-    def gen_ipv4_route(self, nic, gateways, netmask):
-        """
-        Return the routes list needed to configure additional Ipv4 route
-        @return (list): the route list to configure the gateways
-        @param nic (NicBase): the nic to configure
-        @param gateways (str list): the list of gateways
-        """
-        route_list = []
-
-        cidr = ipv4_mask_to_net_prefix(netmask)
-
-        for gateway in gateways:
-            destination = "%s/%d" % (gen_subnet(gateway, netmask), cidr)
-            route_list.append(
-                {
-                    "destination": destination,
-                    "type": "route",
-                    "gateway": gateway,
-                    "metric": 10000,
-                }
-            )
-
-        return route_list
+            dhcp4.update({"dhcp4": False})
+        if dhcp4:
+            return dhcp4
+        else:
+            return None
 
-    def gen_ipv6(self, name, nic):
-        """
-        Return the set of subnets and routes needed to configure the
-        gateways for a nic
-        @return (set): the set of subnets and routes to configure the gateways
-        @param name (str): name of the nic
-        @param nic (NicBase): the nic to configure
-        """
+    def gen_dhcp6(self, nic):
+        if nic.staticIpv6:
+            return {"dhcp6": False}
+        else:
+            # TODO: nic shall explicitly tell it's DHCP6
+            # TODO: set dhcp6-overrides
+            return None
 
-        if not nic.staticIpv6:
-            return ([], [])
+    def gen_addresses(self, nic):
+        address_list = []
+        v4_cidr = 32
+        v6_cidr = 128
 
-        subnet_list = []
+        # Static Ipv4
+        v4_addrs = nic.staticIpv4
+        if v4_addrs:
+            v4 = v4_addrs[0]
+            if v4.netmask:
```

Comment:
Is specifying netmask is optional during static ip setting? please correct me if I'm wrong here.
If there is no netmask given this will result in `<ip>/32` as netmask, is that fine?

I think netmask should be a mandatory field during static ip setting, also please add a test for the same.

Reply:
Per our API definition, netmask is optional for ipv4, if no netmask given, `<ip>/32` is the expected result.

> see subnetMask in https://dp-downloads.broadcom.com/api-content/apis/API_VWSA_001/8.0U3/html/ReferenceGuides/vim.vm.customization.IPSettings.html

but it's mandatory if static ipv6 address included.

> see subnetMask in https://dp-downloads.broadcom.com/api-content/apis/API_VWSA_001/8.0U3/html/ReferenceGuides/vim.vm.customization.FixedIpV6.html

I added 2 tests for both ipv4 and ipv6 cases.